langserver/symbol.go

```diff
@@ -488,3 +479,91 @@ var symbolCacheTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
 func init() {
 	prometheus.MustRegister(symbolCacheTotal)
 }
+
+func listPkgs(ctxt *build.Context, dir string) []string {
+	ch := make(chan string)
+
+	var wg sync.WaitGroup
+	for _, root := range ctxt.SrcDirs() {
+		root := root
+		wg.Add(1)
+		go func() {
+			allPackages(ctxt, root, dir, ch)
+			wg.Done()
+		}()
+	}
+	go func() {
+		wg.Wait()
+		close(ch)
+	}()
+
+	var pkgs []string
+	for p := range ch {
+		pkgs = append(pkgs, p)
+	}
+	sort.Strings(pkgs)
+	return pkgs
+}
+
+// We use a process-wide counting semaphore to limit
+// the number of parallel calls to ReadDir.
+var ioLimit = make(chan bool, 20)
+
+// allPackages is from tools/go/buildutil. We don't use the exported method
+// since it doesn't allow searching from a directory. We need from a specific
+// directory for performance on large GOPATHs.
+func allPackages(ctxt *build.Context, root, start string, ch chan<- string) {
+	root = filepath.Clean(root) + string(os.PathSeparator)
+	start = filepath.Clean(start) + string(os.PathSeparator)
+
+	if strings.HasPrefix(root, start) {
+		// If we are a child of start, we can just start at the
+		// root. A concrete example of this happening is when
+		// root=/goroot/src and start=/goroot
+		start = root
+	}
+
+	if !strings.HasPrefix(start, root) {
```

Comment:
The behaviour of PathHasPrefix is slightly different to what we are doing here. Given we already need the cleaned up root and start above, I think this code is clearer since when reading it we know exactly what strings.HasPrefix does.